cfut/__init__.py

```diff
@@ -41,8 +47,7 @@ def __init__(self, callback, interval=1):
 
     def stop(self):
         """Stop the thread soon."""
-        with self.lock:
-            self.shutdown = True
+        self.shutdown = True
```

Comment:
This may be a dumb question: is it safe to eliminate the lock here? `self.shutdown` is still being concurrently modified and read by two different threads, so there is a data race. Maybe this is OK because of the GIL (i.e., writing to the field and reading from it are guaranteed to be atomic because they are single bytecode operations), but it nonetheless makes me a little nervous.

Reply:
I'm pretty confident this is OK. My understanding is that the GIL guarantees that fundamental Python operations happen atomically, and setting an attribute is very fundamental. The glossary entry says:

> This simplifies the CPython implementation by making the object model (including critical built-in types such as [dict](https://docs.python.org/3/library/stdtypes.html#dict)) implicitly safe against concurrent access.

(Typical attribute access like this is based on dict, stored as `obj.__dict__`)

I think we could even get away without it for the `self.waiting` dict, so long as filenames are always unique, but if for any reason a filename was reused, it could get very confusing, so I'll keep it there. And actually, I realise that the extra Slurm checks should also be inside the lock.

Reply:
Sounds reasonable enough to me!